### PR TITLE
cardano-cli: Add friendly printing of script data for tx outputs (Alonzo)

### PR DESCRIPTION
**Pull request by cblp**

### CODE REVIEW

File: cardano-api/src/Cardano/Api/ScriptData.hs [hunk 1]
```diff
@@ -383,11 +383,14 @@ scriptDataToJsonNoSchema = conv
...
 
     -- Script data allows any value as a key, not just string as JSON does.
-    -- For simple types we just convert them to string dirctly.
+    -- For simple types we just convert them to string directly.
     -- For structured keys we render them as JSON and use that as the string.
     convKey :: ScriptData -> Text
     convKey (ScriptDataNumber n) = Text.pack (show n)
-    convKey (ScriptDataBytes bs) = bytesPrefix
+    convKey (ScriptDataBytes bs)
+      | BSC.all (\c -> Char.isAscii c && Char.isPrint c) bs
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/ScriptData.hs:391]: Why this change?

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]

File: cardano-api/src/Cardano/Api/ScriptData.hs (hunk 1 above)
Comment (cblp) [reply cardano-api/src/Cardano/Api/ScriptData.hs:391]: A map key is often a program identifier, and it's nicer to view it decoded than in hex.

Comment (cblp) [reply cardano-api/src/Cardano/Api/ScriptData.hs:391]: without this,
```json
{"random name": "Vincent Ford"}
```
looks
```json
{"0x72616e646f6d206e616d65": "Vincent Ford"}
```

Comment (cblp) [reply cardano-api/src/Cardano/Api/ScriptData.hs:391]: I followed the comment "For simple types we just convert them to string directly."

Comment (Jimbo4350) [reply cardano-api/src/Cardano/Api/ScriptData.hs:391]: You missed: 
> -- * JSON hex strings with \"0x\" prefix represented as CBOR byte strings
We can't render `ScriptDataBytes` as ASCII because when parsing back to `ScriptData` the prefix `0x` followed by hex indicates that we want to encode this on chain as a CBOR bytestring.

Comment (Jimbo4350) [reply cardano-api/src/Cardano/Api/ScriptData.hs:391]: Have a look at `doc/reference/tx-metadata.md`

Comment (cblp) [reply cardano-api/src/Cardano/Api/ScriptData.hs:391]: What about strings that don't start with `0x`?

Is there another tool to convert a JSON-like datum to corresponding JSON?

Comment (cblp) [reply cardano-api/src/Cardano/Api/ScriptData.hs:391]: So, if a txout datum input is
```js
{"random name": "Vincent Ford"}
```
then inside the tx we have
```hs
txOutDatum = ScriptDataMap [(ScriptDataBytes "random name",ScriptDataBytes "Vincent Ford")]
```

Both key and value are ScriptDataBytes, so they they both must be displayed with no schema as 0x-base-16 strings?

Comment (cblp) [reply cardano-api/src/Cardano/Api/ScriptData.hs]: So, since this contradicts tx-metadata.md, I removed such behaviour from `scriptDataToJsonNoSchema`.

File: cardano-api/src/Cardano/Api/ScriptData.hs [hunk 2]
```diff
@@ -362,11 +361,6 @@ scriptDataToJsonNoSchema = conv
     conv :: ScriptData -> Aeson.Value
     conv (ScriptDataNumber n) = Aeson.Number (fromInteger n)
     conv (ScriptDataBytes bs)
-      | Right s <- Text.decodeUtf8' bs
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/ScriptData.hs:365]: Why? This is plutus script data not tx metadata.

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs [hunk 3]
```diff
@@ -149,29 +153,52 @@ friendlyTxOut (TxOut addr amount mdatum) =
...
-   renderDatum TxOutDatumNone = Aeson.Null
-   renderDatum (TxOutDatumHash _ h) =
-     Aeson.String $ serialiseToRawBytesHexText h
-   renderDatum (TxOutDatum _ sData) =
-     scriptDataToJson ScriptDataJsonDetailedSchema sData
+
+    -- | 'ScriptData' does not support text, only byte strings.
+    -- 'ScriptData.scriptDataToJson' formats byte strings as Base16 blobs.
+    -- So here we try to display text data where byte strings look like text.
+    conv :: ScriptData -> Aeson.Value
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Run/Friendly.hs:184]: There is a lot of repetition of code here from `scriptDataToJson`. I suggest using `scriptDataToJson`, pattern matching on the `Value`s of interest (`Object` in your case) and dealing with the Map keys as you see fit.

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs (hunk 3 above)
Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Run/Friendly.hs:184]: That will be an extra 0x-base-16 encoding followed by extra 0x-base-16 decoding. Is it better?

File: cardano-api/src/Cardano/Api/ScriptData.hs (hunk 2 above)
Comment (cblp) [reply cardano-api/src/Cardano/Api/ScriptData.hs:365]: You sent me tx-metadata docs before. I supposed, tx-metadata rules apply here. Don't they?

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs (hunk 3 above)
Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Run/Friendly.hs:184]: I don't understand what you mean. Can you clarify?

Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Run/Friendly.hs:184]: I don't understand what you mean. Can you clarify?

Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Run/Friendly.hs:184]: So this is what I have in mind:
```
conv :: ScriptData -> ScriptDataJsonSchema -> Aeson.Value
conv sd schema = 
  case scriptDataToJson schema sd of 
    Aeson.Array a -> ...
    ...
```

**Comment by cblp** [comment]
@Jimbo4350 I've removed changes in ScriptData.hs, which don't add to the purpose of the PR.
